gitlab/main.ml

```diff
@@ -0,0 +1,175 @@
+      | `Failed -> { acc with failed = acc.failed + 1 }
+      | `Passed -> { acc with ok = acc.ok + 1 }
+      | `Not_started | `Pending -> { acc with active = acc.active + 1 }
+
+  let count_owner owner (acc : stats) =
+    Index.Repo_set.fold (count_repo ~owner) (Index.get_active_repos ~owner) acc
+
+  let update () =
+    let owners = Index.get_active_owners () in
+    let { ok; failed; active } = Index.Owner_set.fold count_owner owners { ok = 0; failed = 0; active = 0 } in
+    Gauge.set (master "ok") (float_of_int ok);
+    Gauge.set (master "failed") (float_of_int failed);
+    Gauge.set (master "active") (float_of_int active)
+end
+
+(* TODO Isolate gitlab specific configuration from general service config. *)
+module Conf = Ocaml_ci_service.Conf
+
+let setup_log default_level =
+  Prometheus_unix.Logging.init ?default_level ();
+  Mirage_crypto_rng_unix.initialize ();
+  Prometheus.CollectorRegistry.(register_pre_collect default) Metrics.update;
+  match Conf.ci_profile with
+  | `Production -> Logs.info (fun f -> f "Using production configuration")
+  | `Dev -> Logs.info (fun f -> f "Using dev configuration")
+
+let or_die = function
+  | Ok x -> x
+  | Error `Msg m -> failwith m
+
+let run_capnp capnp_public_address capnp_listen_address =
+  match capnp_public_address, capnp_listen_address with
+  | None, None -> Lwt.return (Capnp_rpc_unix.client_only_vat (), None)
+  | Some _, None  ->
+        Lwt.fail_invalid_arg "Public address for Cap'n Proto RPC can't be set without setting a capnp-listen-address to listen on."
+  | Some _, Some _ | None, Some _ ->
+    let listen_address =
+        match capnp_listen_address with
+        | Some listen_address -> listen_address
+        | None -> (Capnp_rpc_unix.Network.Location.tcp ~host:"0.0.0.0" ~port:Conf.Capnp.internal_port)
+    in
+    let public_address =
+        match capnp_public_address with
+        | None -> listen_address
+        | Some public_address -> public_address
+    in
+    let config =
+      Capnp_rpc_unix.Vat_config.create
+        ~public_address
+        ~secret_key:(`File Conf.Capnp.secret_key)
+        listen_address
+    in
+    let rpc_engine, rpc_engine_resolver = Capability.promise () in
+    let service_id = Capnp_rpc_unix.Vat_config.derived_id config "ci" in
+    let restore = Capnp_rpc_net.Restorer.single service_id rpc_engine in
+    Capnp_rpc_unix.serve config ~restore >>= fun vat ->
+    Capnp_rpc_unix.Cap_file.save_service vat service_id Conf.Capnp.cap_file |> or_die;
+    Logs.app (fun f -> f "Wrote capability reference to %S" Conf.Capnp.cap_file);
+    Lwt.return (vat, Some rpc_engine_resolver)
+
+module Gitlab = struct
+  (* Access control policy. *)
+  let has_role user role =
+    match user with
+    | None -> role = `Viewer              (* Unauthenticated users can only look at things. *)
+    | Some user ->
+      match Current_web.User.id user, role with
+      | "gitlab:tmcgilchrist", _ -> true  (* This user has all roles *)
```

Comment:
Figuring this is interim? :)

Reply:
It is equivalent to this module for GitHub https://github.com/ocurrent/ocaml-ci/blob/master/service/github.ml#L6-L11
The main use is for the permissions checks when triggering certain actions in the web UI. That all might change with the Dream redesign.